PCAL64xx driver state retention

benediktibk opened the pull request:
### Problem Statement

For some uses cases a port expander has to keep its outputs in the same state whilst the MCU is switching back and forth between a bootloader and the actual firmware.

### Proposed Solution

The retention subsystem can already be used to share information between a bootloader and firmware, in both directions. I'm using this to store the state of the port expander and load it again during the initialization of the driver. In this specific case, where all registers can also be read, it is actually only necessary to store if the port expander has already been initialized. The actual values can then be read from the registers.
But this approach can be used in a more general way, as the content of the retention area is solely defined by the driver. The size is defined externally, but there is a helper macro to assert during build time that the area is sufficiently large.
During a previous discussion (#62293) the usage of a no-reset flag was also discussed, but this would then only have been a one-way option (bootloader to firmware OR vice versa), and it would have stripped the receiving application from the possibilty to modify the state of the port expander again.

### Problems

There is definitely an issue with changes in the layout how the state is stored. If there is a mismatch between bootloader and firmware this approach will be broken. Luckily, only the initialization state has to be stored for this particular IC, therefore I consider this risk as very small. But this will of course become an issue for other, possibly more complex drivers.

benediktibk committed:
retention: add driver state module

Add a driver state module to the retention subsystem. This can be used by
a driver to persist its state to a retention area and therefore skip some
steps of its initialization after a MCU reset.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
tests: subsys: add build_all tests for retention module driver_state

Add build_all tests for the retention module driver_state. This also
illustrates the possible usage of the module in combination with a
driver.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

nordicjm reviewed (commented):
I think this is really cool and like it! But I would propose making it more generic, in dts instead of having the GPIO itself have additional fields (and the GPIO driver doing the retention parts) I would make the subsys/retention/driver_state.c do that, it would need 2 new function pointers added to the GPIO driver API interface to get the state and set the state (which the driver_state.c file could iterate over and call), that could then be expanded to any driver API through the same sort of functions (though because of how the structs are, they would be different offsets, so each driver class would need to run the same functions but in a different #ifdef block)

benediktibk commented:
> I think this is really cool and like it! But I would propose making it more generic, in dts instead of having the GPIO itself have additional fields (and the GPIO driver doing the retention parts) I would make the subsys/retention/driver_state.c do that, it would need 2 new function pointers added to the GPIO driver API interface to get the state and set the state (which the driver_state.c file could iterate over and call), that could then be expanded to any driver API through the same sort of functions (though because of how the structs are, they would be different offsets, so each driver class would need to run the same functions but in a different #ifdef block)

I'm not sure if I understood the details completely, but I also like the general idea. The big question for me would be if such a change would be acceptable in the GPIO subsystem. Maybe @henrikbrixandersen could give us his opinion on this?

benediktibk commented:
@nordicjm I see an issue with shifting more logic into the retention module. It increases the risk of having a different data layout between bootloader and firmware. As long as the driver itself is responsible for this every IC which has all the registers read-accessible can store only one boolean, and therefore avoid this issue completely.

nordicjm commented:
I created a bit of a knock up diagram here:
![Untitled Diagram drawio](https://github.com/zephyrproject-rtos/zephyr/assets/40387179/f9ad3d55-1345-41ca-831a-93b6bea33685)

So essentially a driver class would need some updates to support it, there would need to be a load state and save state function (the output/input of this data is completely up to the end driver, there should be no assumptions on the size or format of it, though the size should be known so that a retention area of sufficient size can be created), the init function of supported drivers would need to be changed to check if the Kconfig is enabled and if a dts flag is present which determines that initial setup should be skipped, if so then init does not do any setup. Then when the driver state driver initialises, it would check all the registered areas and run the load state function for each, if the data is valid then it will be a pointer to the data, if not then it will just be NULL and the driver should set up the config using the normal default settings. There can be a function for the driver state subsystem that then does the inverse, it calls the save state function for each registered driver and saves the data to the various retention areas so that it can be used for the next application (or reboot, etc.), though there should also be a clear function to clear any driver state that should not be applied for a reboot/next application boot. That's how I envisage it anyhow.

Now there is a drawback to that in that the actual drivers cannot be used until that configuration has been applied, so if you have a GPIO init priority of 40, a sensor init priority of 42 and retention init priority of 50 then the sensor init is going to fail, and I'm not sure if there can be anything done about that in the init level checking script, @fabiobaltieri ?

benediktibk commented:
> I created a bit of a knock up diagram here: 

Pictures help a lot, thanks for that. And with this approach it still would be sufficient to store for instance for the PCAL64xx only one byte, and basically load everything from the registers?

The only issue left is definitely the priority, because if multiple types of drivers would like to use the driver state retention they will likely run into dependency issues, as you already explained. Maybe one could use an instance of the driver state retention per class of drivers (one for GPIOs, one for sensors, ...) and use for each instance the priority of those?
Although, still tricky, because the drivers of each group can also differ. Like the GPIO expanders on I2C busses, which are usually initialized later than the MCU GPIOs ...

fabiobaltieri commented:
> Now there is a drawback to that in that the actual drivers cannot be used until that configuration has been applied, so if you have a GPIO init priority of 40, a sensor init priority of 42 and retention init priority of 50 then the sensor init is going to fail, and I'm not sure if there can be anything done about that in the init level checking script, @fabiobaltieri ?

I suppose you could hack a dependency into it but it sounds like a big mess to be honest, why not initialize the retention subsystem first (not sure what that would imply, load the data from flash or something?) and then let each driver pick up the data and initialize itself in their intended spot in the init sequence?

benediktibk commented:
> > Now there is a drawback to that in that the actual drivers cannot be used until that configuration has been applied, so if you have a GPIO init priority of 40, a sensor init priority of 42 and retention init priority of 50 then the sensor init is going to fail, and I'm not sure if there can be anything done about that in the init level checking script, @fabiobaltieri ?
> 
> I suppose you could hack a dependency into it but it sounds like a big mess to be honest, why not initialize the retention subsystem first (not sure what that would imply, load the data from flash or something?) and then let each driver pick up the data and initialize itself in their intended spot in the init sequence?
> 
> Having the system mess up with the init sequence depending on the state sounds like a bit of a nightmare to me (similar comment that I was voicing in the deferred init discussion btw, lose the benefit of validation tooling, subpriority etc).

The question is about who is controlling the process. With my current approach the driver itself is responsible for triggering the usage of the retained state, with @nordicjm version the retention module would trigger it. I like the idea of moving this logic out of the drivers, as this would be to a certain degree a duplication of logic and code. But at the same time I'm not a big fan of tinkering around with the initialization priorities.

Maybe I can simplify the logic in the drivers with some helpers, to reduce the duplication.

fabiobaltieri commented:
Haven't had time to familiarize with the code but at first sight and from your comment I think this should try to hit both points: have the feature be generic and integrated with the GPIO APIs reusing code whenever possible, but also have the relevant code triggered by the device init itself when appropriate so that the init sequence remains  consistent in both cases.

benediktibk committed:
drivers: gpio: improve logging of PCAL64XXA

Improve logging of the driver for the port expander PCAL64XXA.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
drivers: gpio: extract method to apply initial state of PCAL64XXA

Extract method which applies the initial state in the driver
for the port expander PCAL64XXA.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
drivers: gpio: implement driver state retention for PCAL64XXA

Implement the driver state retention for the port expander
PCAL64XXA.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

nordicjm commented:
The most generic approach is what I am aiming for, so if drivers have a "get state" and "set state" function then it means all the saving/restoring to retention (or actually, another system) can be done outside of the driver otherwise every driver would need to be updated to have swathes of the same code inside. Would be open to ears if there is a better way for it.

Hmm I suppose what could be done is a generic driver class is added which has load config, save config, clear config, is valid config function pointers, and have it set globally using a chosen node, then that would have a user dts entry and would need to use that to look up where to load the data from using retention, so that way retention itself wouldn't be loading it. Don't think dependency checking would work on here though because there would be multiple parts in dts that wouldn't be joined up

benediktibk commented:
I have come up with a slightly modified approach, yet untested on hardware (therefore DNM). This one reduces the amount of duplication which is required. I have also extracted the implementation of the driver state retention itself in the driver for the PCAL64XXA to create a show case what this could look like: https://github.com/zephyrproject-rtos/zephyr/pull/71336/commits/af6814ab3a77fc36eba087b11dd03d8efe410c0d

This looks more reasonable to me now, as there is still no dependency issue but still minimal amount of implementation in the driver itself required.

benediktibk commented:
> Haven't had time to familiarize with the code but at first sight and from your comment I think this should try to hit both points: have the feature be generic and integrated with the GPIO APIs reusing code whenever possible, but also have the relevant code triggered by the device init itself when appropriate so that the init sequence remains consistent in both cases.
> 
> At least to have coherent implementations between drivers, having each one doing its own thing sounds like a bad idea.

I'm not sure about making this feature GPIO specific, because I can think of use cases outside GPIOs. Basically every output, like a DAC, might benefit from this feature.

nordicjm commented on drivers/gpio/gpio_pcal64xxa.c:
```diff
@@ -839,10 +960,31 @@ int pcal64xxa_init(const struct device *dev)
+	}
+#endif /* CONFIG_RETENTION_DRIVER_STATE */
+
+	if (use_retained_state) {
+		LOG_DBG("%s: use retained state", dev->name);
+		rc = pcal64xxa_read_state_from_registers(dev);
+		if (rc != 0) {
+			return rc;
+		}
```
This can be in the ifdef above

nordicjm commented on drivers/gpio/gpio_pcal64xxa.c:
```diff
@@ -812,6 +916,22 @@ static int pcal64xxa_apply_initial_state(const struct device *dev)
+	}
+
+	return 0;
+}
```
can be #ifdef'd

benediktibk replied:
done

nordicjm commented on drivers/gpio/gpio_pcal64xxa.c:
```diff
@@ -472,6 +479,43 @@ static int pcal6408a_pins_cfg_apply(const struct i2c_dt_spec *i2c,
 	return 0;
 }
 
+static int pcal6408a_pins_cfg_read(const struct i2c_dt_spec *i2c,
```
And this

benediktibk replied:
done

nordicjm commented on drivers/gpio/gpio_pcal64xxa.c:
```diff
@@ -622,6 +667,64 @@ static int pcal6416a_pins_cfg_apply(const struct i2c_dt_spec *i2c,
 	return 0;
 }
 
+static int pcal6416a_pins_cfg_read(const struct i2c_dt_spec *i2c,
```
And this

nordicjm commented on tests/subsys/retention/boards/nucleo_f746zg.overlay:
```diff
@@ -0,0 +1,47 @@
+			status = "okay";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			reg = <0x0 0x8>;
```
8 bytes allocated but only 3 are used

nordicjm commented on tests/subsys/retention/boards/nucleo_f746zg.overlay:
```diff
@@ -0,0 +1,47 @@
+	};
+};
+
+&dtcm {
```
Not sure where dtcm is defined but generally you allocate a small portion for retained memory usage e.g.:
```
        sram@2003FC00 {
                compatible = "zephyr,memory-region", "mmio-sram";
                reg = <0x2003FC00 DT_SIZE_K(1)>;
                zephyr,memory-region = "RetainedMem";
                status = "okay";

                retainedmem {
                        compatible = "zephyr,retained-ram";
                        status = "okay";
                        #address-cells = <1>;
                        #size-cells = <1>;
```

benediktibk replied:
I wanted to keep the test as simple as possible, therefore I've used the whole DTCM for the retention. But I also do not mind cutting some space of at the end for this purpose.

nordicjm reviewed (commented):
Ah I see I misunderstood the original proposal, I thought it was to store the GPIO configuration state itself into RAM, not a flag if it then use to read back the state

benediktibk commented:
> Ah I see I misunderstood the original proposal, I thought it was to store the GPIO configuration state itself into RAM, not a flag if it then use to read back the state

You didn't misunderstand it completely, the goal is actually to retain the GPIO configuration. But to avoid possible mismatches in the structure I think it is usually better (if possible, depends on the IC) to store only a flag and load the config from the IC itself.

benediktibk commented on dts/bindings/retention/zephyr,retention-driver-state.yaml:
```diff
@@ -0,0 +1,51 @@
+      reg = <0x0 0x1>;
+    };
+
+    driver_state2: driver_state@0 {
```
done

benediktibk commented on dts/bindings/retention/zephyr,retention-driver-state.yaml:
```diff
@@ -0,0 +1,51 @@
+    driver_state2: driver_state@0 {
+      compatible = "zephyr,retention-driver-state";
+      status = "okay";
+      reg = <0x0 0x8>;
```
done

benediktibk commented on subsys/retention/Kconfig.driver_state:
```diff
@@ -0,0 +1,21 @@
+
+config RETENTION_DRIVER_STATE_INIT_PRIORITY
+	int "Driver state retention init priority"
+	default 87
```
done

benediktibk commented on subsys/retention/driver_state.c:
```diff
@@ -0,0 +1,149 @@
+	} else {
+		LOG_DBG("%s: found invalid content in retention area", dev->name);
+
+		result = retention_write(config->parent, 0, (uint8_t *)&header, sizeof(header));
```
done

nordicjm commented on dts/bindings/retention/zephyr,retention-driver-state.yaml:
```diff
@@ -0,0 +1,51 @@
+  drivers.
+
+  retention0: retention@0 {
+    compatible = "zephyr,retention";
+    status = "okay";
+    reg = <0x0 0x40>;
+    prefix = [04 fa];
+    checksum = <1>;
```
This would not work with the state code, the state code reads and writes the state from 0 and ignores the reg address, this means if you have one retention area and 4 driver state retention parts, at offsets 0x0, 0x1, 0x2 and 0x3 (ignoring checksums and prefixes), then all of them will actually read and write to 0x0. If you wanted to supported partitioning in the driver state code then you would need to use the reg address and pass that to the retention read/write function

nordicjm reviewed (approved):
Looks okay, thanks

fabiobaltieri commented on drivers/gpio/gpio_pcal64xxa.c:
```diff
@@ -891,6 +1047,16 @@ int pcal64xxa_init(const struct device *dev)
 
+#ifdef CONFIG_RETENTION_DRIVER_STATE
+	if (drv_cfg->retention_driver_state != NULL) {
+		rc = retention_driver_state_write(drv_cfg->retention_driver_state, NULL, 0);
```
so the mechanism here is to use `retention_driver_state_write` with no data just for the `is_valid` bit of information, and then that's used to initialize back the state from the gpio registers itself, right?

decsny commented:
The DT property added here seems like a SW configuration not belonging in DT, why does it need to be in DT, instead of some API where the application allocates a space for a device to store state

benediktibk commented:
> The DT property added here seems like a SW configuration not belonging in DT, why does it need to be in DT, instead of some API where the application allocates a space for a device to store state

I suppose you are referring to retention-driver-state? Isn't this somewhat similar to for instance assigning the flash partition via zephyr,flash?
https://github.com/zephyrproject-rtos/zephyr/blob/eddb3822d2aa746026837252b3e337b0c4ef067a/boards/96boards/aerocore2/96b_aerocore2.dts#L20
To me it feels like a hardware configuration, because it specifies in which RAM area the data should be located in.

gmarull reviewed (changes requested):
yet again another misuse of devicetree to structure software configuration.

benediktibk commented:
@gmarull @decsny Is the only issue the property retention-driver-state, or in general the segmentation of the retention area in the device tree?

decsny commented:
> Is the only issue the property retention-driver-state, or in general the segmentation of the retention area in the device tree?

I am not familiar with this retention subsystems (which this seems like the 3rd one now? bit confused why do we have so many layers of retention subsystems, but this is a statement of my ignorance rather than an opinion), but it looks like this is not aligned with the purpose of DT... so this has the Arch WG label on it, which I would have added myself if someone didn't already, because I think it needs more eyes and opinions

carlescufi commented:
> If the consensus ends up being that it doesnt belong in DT like this, I think its fair to say there is a broader problem here that zephyr is trying to be something with all this configuration ability but doesn't actually provide a way to configure at the level of granularity that contributors and users actually need.... Kconfig and DT (if DT is used according to the philosophy of only describing the hardware and matching linux) simply _do not_ provide what people need (even though it feels very natural to try to put these things in DT since technically you can)

I completely agree, and we need to fix this. This is related as well to:
https://github.com/zephyrproject-rtos/zephyr/pull/68127

carlescufi commented:
Architecture WG:

- @henrikbrixandersen considers this a violation of dependencies between drivers. Suggests adding a flag at compile time, but @benediktibk counters that this would not work because the problem is that the application needs to jump back to the bootloader. This means that the bootloader cannot know if the device has been already initialized or not
- @bjarki-trackunit suggests "simplifying" the Devicetree layout to avoid having additional nodes below the retention one. Having N `retention:N retention@N` nodes would be the response to multi-instance
- @fabiobaltieri suggests invoking a special device-specific reset procedure call only when booting from cold in the bootloader. Then the bootloader would use the retention subsystem to detect if it's coming from a cold reset
-

benediktibk commented:
The result of the architecture workgroup meeting was one of two possible options:

1. Create a driver specific (PCAL64xx) reset function call, which is publicly accessible. This would move the responsibility for calling the reset up to the application.
2. Remove the additional compatible retention-driver-state, use retention directly and implement this as part of the device model

I will close this PR and refer to it later on, either with a PR for option 1 or 2, depending on how much time I can carve out for working on this issue.
